File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.py

class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:

        
        st = [] 
        s = [ e for e in s]
        for i in range(len(s)):

            if s[i] =='(':
                st.append(i)
            if s[i] == ')':

                if st:
                    st.pop()
                else:
                    s[i] = '#'

        while st:
            s[st.pop()] = '#'

        s = ''.join(s)
        s = s.replace('#','')
        return s

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.java
class Solution {
    public String minRemoveToMakeValid(String s) {
        
        
        
        Stack<Integer> st = new Stack<Integer>() ; 
        
        int bal = 0 ; 
        for(int ix = 0 ; ix < s.length() ; ix++){
            
            char e = s.charAt(ix) ;
            if( e =='('){
                bal ++ ; 
                st.add( ix) ;
            }else if( e== ')' ) {
                if( bal > 0 ){
                    st.pop() ; 
                    bal -- ; 
                }else {
                    st.add(ix ) ; 
                }
                
            }
        }
        
       String ans ="" ;
       Set<Integer> set = new HashSet<Integer>(st) ; 
       for(int i= 0 ; i < s.length() ; i++){
            
           if( ! set.contains(i)){
                ans += s.charAt(i);     

           }
           
       }
        
        return ans ; 
        
        
    }
}